File: smash-ranks-android/app/src/main/java/com/garpr/android/misc/RegionManagerImpl.java
package com.garpr.android.misc;

import android.content.Context;
import android.content.ContextWrapper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.garpr.android.models.Region;
import com.garpr.android.preferences.Preference;

import java.lang.ref.WeakReference;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

public class RegionManagerImpl implements RegionManager {

    private static final String TAG = "RegionManagerImpl";

    private final List<WeakReference<OnRegionChangeListener>> mListeners;
    private final Preference<Region> mRegion;
    private final Timber mTimber;


    public RegionManagerImpl(@NonNull final Preference<Region> region,
            @NonNull final Timber timber) {
        mListeners = new LinkedList<>();
        mRegion = region;
        mTimber = timber;
    }

    @Override
    public void addListener(@NonNull final OnRegionChangeListener listener) {
        synchronized (mListeners) {
            boolean addListener = true;
            final Iterator<WeakReference<OnRegionChangeListener>> iterator = mListeners.iterator();

            while (iterator.hasNext()) {
                final WeakReference<OnRegionChangeListener> reference = iterator.next();
                final OnRegionChangeListener item = reference.get();

                if (item == null) {
                    iterator.remove();
                } else if (item == listener) {
                    addListener = false;
                }
            }

            if (addListener) {
                mListeners.add(new WeakReference<>(listener));
            }
        }
    }

    @NonNull
    @Override
    public Region getRegion() {
        final Region region = mRegion.get();

        if (region == null) {
            throw new IllegalStateException("region is null");
        }

        return region;
    }

    @NonNull
    @Override
    public Region getRegion(@Nullable final Context context) {
        if (context instanceof RegionHandle) {
            final Region region = ((RegionHandle) context).getCurrentRegion();

            if (region != null) {
                return region;
            }
        }

        if (context instanceof ContextWrapper) {
            return getRegion(((ContextWrapper) context).getBaseContext());
        }

        return getRegion();
    }

    private void notifyListeners() {
        synchronized (mListeners) {
            final Iterator<WeakReference<OnRegionChangeListener>> iterator = mListeners.iterator();

            while (iterator.hasNext()) {
                final WeakReference<OnRegionChangeListener> reference = iterator.next();
                final OnRegionChangeListener item = reference.get();

                if (item == null) {
                    iterator.remove();
                } else {
                    item.onRegionChange(this);
                }
            }
        }
    }

    @Override
    public void removeListener(@Nullable final OnRegionChangeListener listener) {
        synchronized (mListeners) {
            final Iterator<WeakReference<OnRegionChangeListener>> iterator = mListeners.iterator();

            while (iterator.hasNext()) {
                final WeakReference<OnRegionChangeListener> reference = iterator.next();
                final OnRegionChangeListener item = reference.get();

                if (item == null || item == listener) {
                    iterator.remove();
                }
            }
        }
    }

    @Override
    public void setRegion(@NonNull final Region region) {
        mTimber.d(TAG, "old region is \"" + getRegion() + "\", new region is \"" + region + "\"");
        mRegion.set(region);
        notifyListeners();
    }

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/misc/RegionManagerImpl.kt
package com.garpr.android.misc

import android.content.Context
import android.content.ContextWrapper
import com.garpr.android.misc.RegionManager.OnRegionChangeListener
import com.garpr.android.models.Region
import com.garpr.android.preferences.GeneralPreferenceStore
import com.garpr.android.preferences.RankingsPollingPreferenceStore
import java.lang.ref.WeakReference

class RegionManagerImpl(
        private val generalPreferenceStore: GeneralPreferenceStore,
        private val rankingsPollingPreferenceStore: RankingsPollingPreferenceStore,
        private val timber: Timber
) : RegionManager {

    private val listeners = mutableListOf<WeakReference<OnRegionChangeListener>>()


    companion object {
        private const val TAG = "RegionManagerImpl"
    }

    override fun addListener(listener: OnRegionChangeListener) {
        synchronized (listeners) {
            var addListener = true
            val iterator = listeners.iterator()

            while (iterator.hasNext()) {
                val reference = iterator.next()
                val item = reference.get()

                if (item == null) {
                    iterator.remove()
                } else if (item == listener) {
                    addListener = false
                }
            }

            if (addListener) {
                listeners.add(WeakReference(listener))
            }
        }
    }

    override fun getRegion(context: Context?): Region {
        if (context != null) {
            if (context is RegionManager.RegionHandle) {
                val region = (context as RegionManager.RegionHandle).currentRegion

                if (region != null) {
                    return region
                }
            }

            if (context is ContextWrapper) {
                return getRegion(context.baseContext)
            }
        }

        return generalPreferenceStore.currentRegion.get() ?: throw IllegalStateException(
                "currentRegion preference is null!")
    }

    private fun notifyListeners() {
        synchronized (listeners) {
            val iterator = listeners.iterator()

            while (iterator.hasNext()) {
                val reference = iterator.next()
                val item = reference.get()

                if (item == null) {
                    iterator.remove()
                } else {
                    item.onRegionChange(this)
                }
            }
        }
    }

    override fun removeListener(listener: OnRegionChangeListener?) {
        synchronized (listeners) {
            val iterator = listeners.iterator()

            while (iterator.hasNext()) {
                val reference = iterator.next()
                val item = reference.get()

                if (item == null || item == listener) {
                    iterator.remove()
                }
            }
        }
    }

    override fun setRegion(region: Region) {
        timber.d(TAG, "old region is \"${generalPreferenceStore.currentRegion.get()}\", "
                + "new region is \"$region\"")
        rankingsPollingPreferenceStore.lastPoll.delete()
        rankingsPollingPreferenceStore.rankingsDate.delete()
        generalPreferenceStore.currentRegion.set(region)
        notifyListeners()
    }

}
